lib/inspec/config.rb

```diff
@@ -0,0 +1,286 @@
+      puts
+    end
+
+    #-----------------------------------------------------------------------#
+    #                         Reading Config Files
+    #-----------------------------------------------------------------------#
+
+    # Regardless of our situation, end up with a readable IO object
+    def resolve_cfg_io(cli_opts, cfg_io)
+      unless cfg_io
+        # TODO: deprecate --json-config
+        path = cli_opts[:config] || cli_opts[:json_config]
+        if path == '-'
+          Inspec::Log.warn 'Reading JSON config from standard input' if STDIN.tty?
+          path = STDIN
+        elsif path.nil?
+          default_path = File.join(Inspec.config_dir, 'config.json')
+          path = default_path if File.exist?(default_path)
+        elsif !File.exist?(path)
+          raise ArgumentError, "Could not read configuration file at #{path}"
+        end
+        cfg_io = File.open(path) if path
+        cfg_io ||= StringIO.new('{ "version": "1.1" }')
+      end
+      cfg_io
+    end
+
+    def read_cfg_file_io(cfg_io)
+      contents = cfg_io.read
+      begin
+        @cfg_file_contents = JSON.parse(contents)
+        validate_config_file_contents!
+      rescue JSON::ParserError => e
+        raise Inspec::ConfigError::MalformedJson, "Failed to load JSON configuration: #{e}\nConfig was: #{contents}"
+      end
+      @cfg_file_contents
+    end
+
+    def file_version
+      @cfg_file_contents['version'] || :legacy
+    end
+
+    def legacy_file?
+      file_version == :legacy
+    end
+
+    def config_file_cli_options
+      if legacy_file?
+        # Assume everything in the file is a CLI option
+        @cfg_file_contents
+      else
+        @cfg_file_contents['cli_options'] || {}
+      end
+    end
+
+    def config_file_reporter_options
+      # This is assumed to be top-level in both legacy and 1.1.
+      # Technically, you could sneak it in the 1.1 cli opts area.
+      @cfg_file_contents.key?('reporter') ? { 'reporter' => @cfg_file_contents['reporter'] } : {}
+    end
+
+    #-----------------------------------------------------------------------#
+    #                            Validation
+    #-----------------------------------------------------------------------#
+    def validate_config_file_contents!
+      version = @cfg_file_contents['version']
+
+      # Assume legacy format, which is unconstrained
+      return unless version
+
+      unless version == '1.1'
+        raise Inspec::ConfigError::Invalid, "Unsupported config file version '#{version}' - currently supported versions: 1.1"
+      end
+
+      valid_fields = %w{version cli_options credentials compliance reporter}.sort
+      @cfg_file_contents.keys.each do |seen_field|
+        unless valid_fields.include?(seen_field)
+          raise Inspec::ConfigError::Invalid, "Unrecognized top-level configuration field #{seen_field}.  Recognized fields: #{valid_fields.join(', ')}"
+        end
+      end
+    end
+
+    def validate_reporters(reporters)
```

Comment:
I'm curious how the schema will be validated in future PRs, right now I don't see a way to avoid

```
reporter: "json"
```

Reply:
I'm not sure I follow your concern.  There is no intended change to the logic of parsing or validating the reporters.

Reply:
I was concerned if someone had an invalid config (though valid JSON). For example if someone used just `reporter: "json"` then https://github.com/inspec/inspec/blob/09e0579f5c34b642de939dbfb54aaf220249bf03/lib/inspec/config.rb#L147 would fail with a stacktrace.

Reply:
So, when Inspec::Config finds a problem, it throws an exception.  Things that consume Inspec::Config (none of which are included on this PR) are then responsible for catching exceptions and handling them - which might mean reporting a graceful error to the user, with no stacktrace.

The existing code throws exceptions, as well - see https://github.com/inspec/inspec/blob/feefdcf9afba80f4e666e6dad5ffce6bd30019dd/lib/inspec/base_cli.rb#L171 

The existing code calls validate_reporters through a long chain, eventually surfacing as the `opts` call which is used as one of the first things in nearly every command method.  And each command method provides a facility for handling exceptions - for example, here is `inspec exec`'s exception handler block:

https://github.com/inspec/inspec/blob/master/lib/inspec/cli.rb#L252